ipld/src/selector/walk.rs

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::super::{Error, Ipld, Path, PathSegment};
+use super::Selector;
+use async_recursion::async_recursion;
+use async_trait::async_trait;
+use cid::Cid;
+
+impl Selector {
+    /// Walks all nodes visited (not just matched nodes) and executes callback with progress and
+    /// Ipld node. An optional link loader/ resolver is passed in to be able to traverse links.
+    pub async fn walk_all<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
+        callback: F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld, VisitReason) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        Progress {
+            resolver,
+            path: Path::default(),
+        }
+        .walk_all(ipld, self, &callback)
+        .await
+    }
+
+    /// Walks a graph of Ipld nodes, executing the callback only on the nodes "matched".
+    /// If a resolver is passed in, links will be able to be traversed.
+    pub async fn walk_matching<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
+        callback: F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        self.walk_all(ipld, resolver, |prog, ipld, reason| -> Result<(), String> {
+            if let VisitReason::SelectionMatch = reason {
+                return callback(prog, ipld);
+            }
+            Ok(())
+        })
+        .await
+    }
+}
+
+/// Provides reason for callback in traversal for `walk_all`.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum VisitReason {
+    /// Ipld node visited was a specific match.
+    SelectionMatch,
+    /// Ipld node was visited while searching for matches.
+    SelectionCandidate,
+}
+
+#[async_trait]
+pub trait LinkResolver {
+    #[allow(unused_variables)]
+    /// Resolves a Cid link into it's respective Ipld node, if it exists.
+    async fn load_link(&self, link: &Cid) -> Result<Option<Ipld>, String> {
+        Err("load_link not implemented on the LinkResolver".into())
+    }
```

Comment:
```suggestion
    async fn load_link(&self, link: &Cid) -> Result<Option<Ipld>, String>;
```
Seems to me like you'd want to require conforming types to implement this? They can always return an error themselves if that's appropriate behavior

Reply:
Nah there will be other methods added to this if/when they are needed, makes sense to just default to not implemented so someone implementing their own resolver doesn't have to specify the functions if they don't want to use

Reply:
Should it return an error if you don't implement it though, rather than returning `Ok(None)`?

Reply:
Yeah so I'm not a huge fan of the setup, but the intention is either you don't specify a resolver and those will just be skipped, or you do specify a resolver and if it requires that function for resolving, then it would error. 

Thinking about it more maybe it is just best to return None for loading links, I was just thinking ahead to the other functions like storing links.

Reply:
If quietly skipping links is not deemed acceptable then I would generally speaking rather have the compiler tell me about it than seeing some error at runtime. But I'm having a hard time seeing how this API will evolve exactly so I can't say for sure which approach is the most logical

Reply:
Ah ya mb I forgot about this and had merged in, definitely open to switching it to not default, just for now doesn't matter because this isn't an API used outside of our codebase yet